### PR TITLE
Wok 592 validate iip

### PR DESCRIPTION
**Pull request by VictorHarbo**
Implements validation of IIP parameters.

### CODE REVIEW

File: src/test/java/dk/kb/image/IIPValidationTest.java [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class IIPValidationTest {
```
Comment (tokee) [inline src/test/java/dk/kb/image/IIPValidationTest.java:15]: Overall very thorough checking for problematic values, but you are missing checks for _valid_ values as well as missing (`null`) values.

**Review by tokee — CHANGES_REQUESTED** [review]
One (1) coding error. The rest is only suggestions for adjustments.

File: src/main/java/dk/kb/image/IIPParamValidation.java [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+        if (fif == null || fif.isEmpty()) {
+            throw new InvalidArgumentServiceException("The parameter FIF must be defined");
+        }
+    }
+
+    /**
+     * Validate that CVT is set to either jpeg or png, as these are the allowed types for export in IIP
+     */
+    public static void cvtValidation(String cvt){
+        if (cvt != null) {
```
[the remark this answers is not in the document]
Comment (VictorHarbo) [reply src/main/java/dk/kb/image/IIPParamValidation.java:107]: The null check in line 107 was meant to ensure, that the service doesn't fail, when CVT is not in use. For example when asking for JTL or PTL

File: src/main/java/dk/kb/image/IIPParamValidation.java [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+        }
+    }
+
+    /**
+     * Validate that WID(width) is only set, when used together with CVT.
+     * WID has to be set as width in pixels.
+     * For example 1200
+     */
+    public static void widValidation(Long wid, String cvt) {
+        if (cvt == null || cvt.isEmpty() && wid != null) {
```
[the remark this answers is not in the document]
Comment (VictorHarbo) [reply src/main/java/dk/kb/image/IIPParamValidation.java:178]: `wid` is of type long. I can make this check by doing `!wid.toString().isEmpty()` but I don't know if that would make any sense as I cant see how `wid` can be an empty string in the first place?

File: src/main/java/dk/kb/image/IIPParamValidation.java [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+        }
+    }
+
+    /**
+     * Validate that HEI(height) is only set, when used together with CVT.
+     * HEI has to be set as height in pixels.
+     * For example 800
+     */
+    public static void heiValidation(Long hei, String cvt) {
+        if (cvt == null || cvt.isEmpty() && hei != null) {
```
[the remark this answers is not in the document]
Comment (VictorHarbo) [reply src/main/java/dk/kb/image/IIPParamValidation.java:189]: `hei` is of type long. I can make this check by doing `!hei.toString().isEmpty()` but I don't know if that would make any sense as I cant see how `hei` can be an empty string in the first place?

File: src/test/java/dk/kb/image/IIPValidationTest.java [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+
+        // Creating test values and getting exception messages
+        String expectedXMessage = "The value of x in parameter RGN is out of bounds. It has to be between 0.0 and 1.0";
+        String actualXMessage = exceptionX.getMessage();
+        String expectedYMessage = "The value of y in parameter RGN is out of bounds. It has to be between 0.0 and 1.0";
+        String actualYMessage = exceptionY.getMessage();
+        String expectedWMessage = "The value of w in parameter RGN is out of bounds. It has to be between 0.0 and 1.0";
+        String actualWMessage = exceptionW.getMessage();
+        String expectedHMessage = "The value of h in parameter RGN is out of bounds. It has to be between 0.0 and 1.0";
+        String actualHMessage = exceptionH.getMessage();
```
[the remark this answers is not in the document]
Comment (VictorHarbo) [reply src/test/java/dk/kb/image/IIPValidationTest.java:52]: We didn't really find the perfect solution for this over lunch last week. I will try to implement a check on a higher level.

File: src/main/java/dk/kb/image/IIPParamValidation.java [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+        }
+    }
+
+    /**
+     * Validate that JTL is correctly set, when exporting JPEG tiles.
+     * JTL should contain index n at resolution level r, specified as r,n.
+     * For example: 5,800
+     */
+    public static void jtlValidation(List<Integer> jtl){
+        if (jtl.size() != 0 & jtl != null & !jtl.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (tokee) [reply src/main/java/dk/kb/image/IIPParamValidation.java]: The problem still exists at line 217, where `if ((cvt == null || cvt.isEmpty()) && (!rgn.isEmpty() || rgn != null)) {` will fail with a ǸullPointerException` if `rgn` is `null`.

File: src/main/java/dk/kb/image/IIPParamValidation.java [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+            }
+
+
+            // Slice input string and get indexes of delimiters
+            int endOfR = pfl.indexOf(":");
+            int endOfPairOne = pfl.indexOf("-");
+            String pairOne = pfl.substring(endOfR + 1, endOfPairOne);
+            String pairTwo = pfl.substring(endOfPairOne + 1);
+            int pairOneComma = pairOne.indexOf(",");
+            int pairTwoComma = pairTwo.indexOf(",");
```
[the remark this answers is not in the document]
Comment (tokee) [reply src/main/java/dk/kb/image/IIPParamValidation.java]: Very nice now.

File: src/test/java/dk/kb/image/IIPValidationTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (tokee) [reply src/test/java/dk/kb/image/IIPValidationTest.java:53]: Let's leave them as-is. Changing them now is the same work as changing them later and if the Exceptions are never adjusted there is no need for changing.

**Pull request merged**